### PR TITLE
indents(ecma): use `@auto` for comments

### LINKED ISSUE #2544: Indent in comment is broken for TS and JS files
### Describe the bug

Try typing in text and hit enter in a multiline comment. You would see an issue similar to this:

153990427-6335faf3-c713-4e92-a258-71a6b255254d

I must mention that it works fine in a typescript react (.tsx) file:

anim-rg

### To Reproduce

Use this config

```lua
local ts = require'nvim-treesitter.configs'

ts.setup {
  ensure_installed = "maintained",
  highlight = {
    enable = true,
  },
}
```

and open a JavaScript or a TypeScript file. Try to do a multiline comment. You would notice upon hitting enter after typing text in the first line `*` (star) doesn't align for the next one.

### Expected behavior

Should behave either:
- like what it behaves in a TypeScript React (.tsx) file, or
- the way stock Vim works

### Output of `:checkhealth nvim-treesitter`

```markdown
nvim-treesitter: require("nvim-treesitter.health").check()
========================================================================
## Installation
  - WARNING: `tree-sitter` executable not found (parser generator, only needed for :TSInstallFromGrammar, not required for :TSInstall)
  - OK: `node` found v14.18.3 (only needed for :TSInstallFromGrammar)
  - OK: `git` executable found.
  - OK: `cc` executable found. Selected from { vim.NIL, "cc", "gcc", "clang", "cl", "zig" }
    Version: cc (Ubuntu 10.3.0-1ubuntu1) 10.3.0
  - OK: Neovim was compiled with tree-sitter runtime ABI version 14 (required >=13). Parsers must be compatible with runtime ABI.

## Parser/Features H L F I J
  - hocon          ✓ . . . ✓ 
  - fennel         ✓ ✓ . . ✓ 
  - llvm           ✓ . . . . 
  - surface        ✓ . ✓ ✓ ✓ 
  - glimmer        ✓ . . . . 
  - vue            ✓ . ✓ ✓ ✓ 
  - regex          ✓ . . . . 
  - rasi           ✓ ✓ ✓ ✓ . 
  - lua            ✓ ✓ ✓ ✓ ✓ 
  - glsl           ✓ ✓ ✓ ✓ ✓ 
  - gleam          ✓ ✓ ✓ ✓ ✓ 
  - latex          ✓ . ✓ . ✓ 
  - php            ✓ ✓ ✓ ✓ ✓ 
  - vala           ✓ . . . . 
  - rst            ✓ ✓ . . ✓ 
  - perl           ✓ . . . . 
  - python         ✓ ✓ ✓ ✓ ✓ 
  - java           ✓ ✓ . ✓ ✓ 
  - c_sharp        ✓ ✓ ✓ . ✓ 
  - eex            ✓ . . . ✓ 
  - fusion         ✓ ✓ ✓ ✓ . 
  - pascal         ✓ ✓ ✓ ✓ ✓ 
  - supercollider  ✓ ✓ ✓ ✓ ✓ 
  - yang           ✓ . ✓ . . 
  - yaml           ✓ ✓ ✓ ✓ ✓ 
  - dockerfile     ✓ . . . ✓ 
  - tlaplus        ✓ ✓ ✓ . ✓ 
  - elixir         ✓ ✓ ✓ ✓ ✓ 
  - dot            ✓ . . . ✓ 
  - graphql        ✓ . . ✓ ✓ 
  - cmake          ✓ . ✓ . . 
  - beancount      ✓ . ✓ . . 
  - tsx            ✓ ✓ ✓ ✓ ✓ 
  - r              ✓ ✓ . ✓ . 
  - scala          ✓ . ✓ . ✓ 
  - ninja          ✓ . ✓ ✓ . 
  - lalrpop        ✓ ✓ . . . 
  - norg           . . . . . 
  - css            ✓ . ✓ ✓ ✓ 
  - make           ✓ . . . ✓ 
  - prisma         ✓ . . . . 
  - gowork         ✓ . . . ✓ 
  - http           ✓ . . . ✓ 
  - hjson          ✓ ✓ ✓ ✓ ✓ 
  - pioasm         ✓ . . . ✓ 
  - rust           ✓ ✓ ✓ ✓ ✓ 
  - json5          ✓ . . . ✓ 
  - vim            ✓ ✓ . . ✓ 
  - html           ✓ ✓ ✓ ✓ ✓ 
  - ocaml          ✓ ✓ ✓ . ✓ 
  - clojure        ✓ ✓ ✓ . ✓ 
  - zig            ✓ . ✓ ✓ ✓ 
  - typescript     ✓ ✓ ✓ ✓ ✓ 
  - comment        ✓ . . . . 
  - bibtex         ✓ . ✓ ✓ . 
  - jsdoc          ✓ . . . . 
  - svelte         ✓ . ✓ ✓ ✓ 
  - turtle         ✓ ✓ ✓ ✓ ✓ 
  - hcl            ✓ . ✓ ✓ ✓ 
  - sparql         ✓ ✓ ✓ ✓ ✓ 
  - gdscript       ✓ ✓ . ✓ ✓ 
  - bash           ✓ ✓ ✓ . ✓ 
  - query          ✓ ✓ ✓ ✓ ✓ 
  - heex           ✓ ✓ ✓ ✓ ✓ 
  - nix            ✓ ✓ ✓ . ✓ 
  - json           ✓ ✓ ✓ ✓ . 
  - pug            ✓ . . . ✓ 
  - julia          ✓ ✓ ✓ ✓ ✓ 
  - javascript     ✓ ✓ ✓ ✓ ✓ 
  - jsonc          ✓ ✓ ✓ ✓ ✓ 
  - scss           ✓ . . ✓ . 
  - ql             ✓ ✓ . ✓ ✓ 
  - erlang         . . . . . 
  - c              ✓ ✓ ✓ ✓ ✓ 
  - dart           ✓ ✓ . ✓ ✓ 
  - ruby           ✓ ✓ ✓ ✓ ✓ 
  - toml           ✓ ✓ ✓ ✓ ✓ 
  - cpp            ✓ ✓ ✓ ✓ ✓ 
  - gomod          ✓ . . . ✓ 
  - ledger         ✓ . ✓ ✓ ✓ 
  - go             ✓ ✓ ✓ ✓ ✓ 
  - kotlin         ✓ ✓ ✓ . ✓ 
  - fish           ✓ ✓ ✓ ✓ ✓ 
  - cuda           ✓ ✓ ✓ ✓ ✓ 
  - commonlisp     ✓ ✓ ✓ . . 
  - ocaml_interface✓ ✓ ✓ . ✓ 

  Legend: H[ighlight], L[ocals], F[olds], I[ndents], In[j]ections
         +) multiple parsers found, only one will be used
         x) errors found in the query, try to run :TSUpdate {lang}
```

### Output of `nvim --version`

```text
NVIM v0.7.0-dev+1109-gfaeff49cb
Build type: RelWithDebInfo
LuaJIT 2.1.0-beta3
```

### Additional context

Options I tried:
- [Treesitter's indent option](https://github.com/nvim-treesitter/nvim-treesitter#indentation) — this breaks multiline comment indent in .tsx file too
- The PR which resolves #2515 — it didn't help

There's a related discussion [happened on reddit](https://old.reddit.com/r/neovim/comments/ss8mp7/any_idea_why_when_writing_jsst_multline_comment/) (informal, not with neovim core members)

### PR DESCRIPTION
**Pull request by theHamsta**
Fixes #2544

@ajitid

### CODE REVIEW

**Comment by ajitid** [comment]
Hey @theHamsta, do you have a GIF/screenshot showing that it is working? Because it is still broken for me with this PR. Here's what I did:

1. Replaced

```vim
Plug 'nvim-treesitter/nvim-treesitter', {'do': ':TSUpdate'}
```

with

```vim
Plug 'theHamsta/nvim-treesitter', {'branch': 'ecma-auto-comment', 'do': ':TSUpdate'}
```

2. Just for surety, I did `:TSUninstall all` and restarted nvim so that it could freshly install parsers.

**Comment by ajitid** [comment]
It is not working for me with 54e4530

**Comment by theHamsta** [comment]
It's working for me and it's also working in the tests. Did you restart?

It will not work when starting a comment with `/*` as it's taking it for a `(ERROR (binary_expression (regex)))`

**Comment by ajitid** [comment]
> It will not work when starting a comment with `/*` as it's taking it for a `(ERROR (binary_expression (regex)))`

Just to clarify — you're meaning to say that text shouldn't be written on the line where starting comment string is `/*`, right? If yes, then I'm not writing text there. 

> Did you restart?

I did restart. I'll try to remove everything and check again

**Comment by theHamsta** [comment]
Me and the tests have the behavior of your TSX in JS and TS. The indents are almost the same for all three languages. There has to be some difference in the Vimscript filetype configuration.

**Comment by ajitid** [comment]
@theHamsta I just tried with basically removing everything (here's the [dotfile I used](https://gist.github.com/ajitid/d80093d68962cc67e0c7659d0eaffcc7) for that matter) and I can still see the same issue.

**Comment by theHamsta** [comment]
But you're not using our indent module in your config https://github.com/nvim-treesitter/nvim-treesitter/blob/2298a7584414af40a6a09cbe72a81175382992c7/scripts/minimal_init.lua#L12 but because we disable `syntax` it's possible that we break one of the traditional indent methods

**Comment by ajitid** [comment]
Oh, my mistake! I tried, now it works for all files except there's  a slight issue in `.ts` ones. In them, I had manually align ending multiline comment (`*/`):

anim-rg

**Comment by theHamsta** [comment]
Yes, this is the issue I was speaking of before. At the moment when we have just `/*` it's a regex in a error node for tree-sitter. I tried to match it but until now without success. Maybe we could use `@auto` always on errors.

**Comment by ajitid** [comment]
Oh okay. Weird that it only occurs in `.ts` but not in `.js` or `.tsx` files, isn't it? Anyway, I'm quite happy as we've made a significant progress here.

> Maybe we could use @auto always on errors.

Totally on you if you want to merge the PR and close the issue or to give another stab fixing it. Thanks for your efforts!

**Comment by theHamsta** [comment]
@ajitid  can you check again? The behavior for unfinished `/*` should be fixed now.

**Comment by theHamsta** [comment]
No idea why tests are failing without error. Seems similar to #2291

**Comment by ajitid** [comment]
Unfortunately, it's still working [as before](https://github.com/nvim-treesitter/nvim-treesitter/pull/2545#issuecomment-1045781362) for me and new issues have popped up now:

- This is a .tsx file. If a multiline comment is opened even a level deep, it breaks:

anim-rg

**Comment by ajitid** [comment]
Here's what I got (all of these GIFs are of a `.tsx` file):

VS Code's multiline comment:

anim-rg

vs this PRs multiline comment:

anim-rg

vs what stock Vim gives us:

anim-rg

Stock Vim behaves in an odd fashion, so let's keep it aside for now. But this PR has different indent behavior when a comment is made at the first column vs when it is done at an indent.

Here's VSCode's JSDoc comment:

anim-rg

To me VSCode's behavior seem to be the most appropriate here.

**Comment by theHamsta** [comment]
Hmm, apparently it depends on the context on how the `ERROR` because it will add anything before to the error so that our regex will not match. I will try multi-line Lua regexes.

**Comment by ajitid** [comment]
> I find your GIFs super unhelpful. It's difficult to understand at what moment something should behave differently. It's usually better to have plain text with a marker where something should be inserted and what the result after indentation will be. Plus I can copy+paste your complete failure case to make it a test.

Yep noted. Will use text from now on.

**Comment by theHamsta** [comment]
@ajitid I didn't mean in any way to criticize you. I'm very glad that you discovered that bug also in the "fixes". We can just save us some round-trips when it's clear what we mean. Best to communicate is failing example or even better a test for that.

**Review by theHamsta — COMMENTED** [review]
todo: fix regex

**Comment by icholy** [comment]
I just tested this branch and it still seems broken to me.



If the comment isn't closed, the next line does not get indented correctly.

``` ts
/** <- hit [ENTER] here
*
```

If the comment is closed, the next line gets indented correctly.

``` ts
/** <- hit [ENTER] here
 *
 */
```

**Comment by khamer** [comment]
I ran into this, switched to @theHamsta 's fork and it resolved the issues for me. I know it's nearly a thousand commits behind master now; what's the status here? Anything I can do to help? Any chance of a merge? At worst, I might see about rebasing the fork on top of master for my own use.

**Comment by theHamsta** [comment]
I don't remember whether this was an improvement or regression. Since the tests pass I guess it's ok? Can other JS please comment? I can maybe give it a try to solve the remaining issues that we still had here.

**Comment by khamer** [comment]
I'm not sure - the comment indentation has been wrong since I started using this, so it might be an improvement. Compared to how C-style block comment indentation is handled w/o treesitter, it's a regression.

**Comment by theHamsta** [comment]
I think we should merge. It is the same how comments are currently handled in C/C++ and Java. Though Go still uses the `@ignore` way.

File: queries/ecma/indents.scm [hunk 1]
```diff
@@ -47,9 +47,8 @@
...
-[
-  (comment)
-  (template_string)
-] @ignore
-
 (ERROR) @auto
+(template_string) @ignore
+(comment) @auto
+;; Probably, also a comment but unfinished
+((ERROR) @auto (#lua-match? @auto "/%*"))
```
[the remark this answers is not in the document]
Comment (amaanq) [reply queries/ecma/indents.scm:54]: ping @theHamsta this would be really nice to merge :)

**Comment by clason** [comment]
Needs to be adapted to the new upstream format, if still desired.

**Pull request closed without merging**